flink-libraries/flink-table/src/main/scala/org/apache/flink/table/codegen/CodeGenerator.scala

```diff
@@ -1463,21 +1465,23 @@ class CodeGenerator(
     */
   def addReusableFunction(function: UserDefinedFunction): String = {
     val classQualifier = function.getClass.getCanonicalName
-    val fieldTerm = s"function_${classQualifier.replace('.', '$')}"
+    val functionSerializedData = serialize(function)
+    val fieldTerm =
+      s"""
+         |function_${classQualifier.replace('.', '$')}_${DigestUtils.md5Hex(functionSerializedData)}
```

Comment:
Hi @wuchong 
No, scalar UDF in Scala Table API not works well.  
The reason is that when we create `ScalarSqlFunction`, we apply` scalarFunction.getClass.getCanonicalName` as sql identifier, which produces the wrong result.

Hi, @fhueske  
So far, we have discussed a lot of UDF implementations in this PR, so I agree with merge FLINK-5794 into this PR.